File: SecondYear/Junior-Hawk_ArchitAgarwal_2226cseai1186_2/Week_2/parenthesis.cpp
void generateParenthesis(int left, int right, string current,vector<string>& result) {
    if (left == 0 && right == 0) {
        result.push_back(current);
        return;
    }

    if (left > 0) {
        generateParenthesis(left - 1, right, current + '(', result);
    }

    if (right > left) {
        generateParenthesis(left, right - 1, current + ')', result);
    }
}
vector<string> validParenthesis(int n){
    vector<string> result;
    generateParenthesis(n, n, "", result);
    return result;
}

File: SecondYear/gamandeepsingh_GamandeepSingh_2226csit1042_2/Week_2/Parenthesis.java
import java.util.ArrayList;

public class Solution {
    public static ArrayList<String> validParenthesis(int n) {
        // Write your code here.
        ArrayList<String> result = new ArrayList<>();
        generateParenthesis(n, n, "", result);
        return result;
    }

    private static void generateParenthesis(int left, int right, String current, ArrayList<String> result) {
        if (left == 0 && right == 0) {
            result.add(current);
            return;
        }
        if (left > 0) {
            generateParenthesis(left - 1, right, current + "(", result);
        }
        if (right > left) {
            generateParenthesis(left, right - 1, current + ")", result);
        }
    }
}
